include/ensmallen_bits/nsgaiii/nsgaiii_impl.hpp

```diff
@@ -0,0 +1,320 @@
+/**
+ * @file nsgaiii_impl.hpp
+ * @author Rahul Ganesh Prabhu
+ *
+ * Non-dominated Sorting Genetic Algorithm - III (NSGA-III)
+ * A multi-objective optimizer. \temp
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_NSGAIII_NSGAIII_IMPL_HPP
+#define ENSMALLEN_NSGAIII_NSGAIII_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "nsgaiii.hpp"
+
+namespace ens {
+
+inline NSGAIII::NSGAIII(const size_t populationSize,
+					    const size_t maxGenerations,
+					    const double crossoverProb,
+					    const double distrIndex) :
+		populationSize(populationSize),
+		maxGenerations(maxGenerations),
+		crossoverProb(crossoverProb),
+		distrIndex(distrIndex),
+		userDefinedSet(false)
+{ /* Nothing to do here */ }
+
+template<typename MultiObjectiveFunctionType>
+arma::cube NSGAIII::Optimize(MultiObjectiveFunctionType& function, arma::mat& iterate)
+{
+  // The user didn't define a reference set.
+  if (!userDefinedSet)
+  {
+
+  }
+
+  // Initialize initial population.
+  arma::cube population = arma::randn(iterate.n_rows, iterate.n_cols, populationSize);
+  population.each_slice() += iterate;
+
+  // Initialize helper variables.
+  std::vector<size_t> S;
+  arma::cube R, Q;
+  std::vector<std::vector<size_t>> fronts;
+
+  for (size_t gen = 0; gen < maxGenerations; gen++)
+  {
+	// Mate and create offsprings.
+	Q = Mate(population);
+
+	// Create a population made of the old population and the new offsprings.
+	R = arma::join_slices(population, Q);
+
+	// Evaluate R.
+	arma::mat fitnessValues(function.NumObjectives(), R.n_slices);
```

Comment:
Looks like we could put this outside the for loop and reuse the parameter for each iteration.